### PR TITLE
Sapling validation

### PR DESCRIPTION
**Pull request by ebfull**
**Also review https://github.com/zcash/librustzcash/pull/7**

This is an attempt to tackle much of https://github.com/zcash/zcash/issues/3065

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -967,6 +967,67 @@ bool ContextualCheckTransaction(const CTransaction& tx, CValidationState &state,
...
+        try {
+            sighash_value = SignatureHash(scriptCode, tx, NOT_AN_INPUT, SIGHASH_ALL, 0, consensusBranchId);
+        } catch (std::logic_error ex) {
+            return state.DoS(100, error("ContextualCheckTransaction(): error computing signature hash"),
+                                REJECT_INVALID, "error-computing-signature-hash");
+        }
+
+        auto ctx = librustzcash_sapling_verification_ctx_init();
+
+        BOOST_FOREACH(const SpendDescription &spend, tx.vShieldedSpend) {
```
Comment (str4d) [inline src/main.cpp]: Use `for (const SpendDescription &spend : tx.vShieldedSpend) {`

File: src/main.cpp [hunk 2]
```diff
@@ -967,6 +967,67 @@ bool ContextualCheckTransaction(const CTransaction& tx, CValidationState &state,
...
+                &spend.spendAuthSig[0],
+                sighash_value.begin()
+            ))
+            {
+                return state.DoS(100, error("ContextualCheckTransaction(): Sapling spend description invalid"),
+                                      REJECT_INVALID, "bad-txns-sapling-spend-description-invalid");
+            }
+        }
+
+        BOOST_FOREACH(const OutputDescription &output, tx.vShieldedOutput) {
```
Comment (str4d) [inline src/main.cpp]: Use `for (const OutputDescription &output : tx.vShieldedOutput) {`

File: src/main.cpp [hunk 3]
```diff
@@ -967,6 +967,67 @@ bool ContextualCheckTransaction(const CTransaction& tx, CValidationState &state,
...
+        auto ctx = librustzcash_sapling_verification_ctx_init();
+
+        BOOST_FOREACH(const SpendDescription &spend, tx.vShieldedSpend) {
+            if (!librustzcash_sapling_check_spend(
+                ctx,
+                spend.cv.begin(),
+                spend.anchor.begin(),
+                spend.nullifier.begin(),
+                spend.rk.begin(),
+                &spend.zkproof[0],
```
Comment (str4d) [inline src/main.cpp]: `spend.zkproof.begin(),`

File: src/main.cpp [hunk 4]
```diff
@@ -967,6 +967,67 @@ bool ContextualCheckTransaction(const CTransaction& tx, CValidationState &state,
...
+                &spend.spendAuthSig[0],
```
Comment (str4d) [inline src/main.cpp]: `spend.spendAuthSig.begin(),`

File: src/main.cpp [hunk 5]
```diff
@@ -967,6 +967,67 @@ bool ContextualCheckTransaction(const CTransaction& tx, CValidationState &state,
...
+            }
+        }
+
+        BOOST_FOREACH(const OutputDescription &output, tx.vShieldedOutput) {
+            if (!librustzcash_sapling_check_output(
+                ctx,
+                output.cv.begin(),
+                output.cm.begin(),
+                output.ephemeralKey.begin(),
+                &output.zkproof[0]
```
Comment (str4d) [inline src/main.cpp]: `output.zkproof.begin()`

File: src/main.cpp [hunk 6]
```diff
@@ -967,6 +967,67 @@ bool ContextualCheckTransaction(const CTransaction& tx, CValidationState &state,
...
+            {
+                return state.DoS(100, error("ContextualCheckTransaction(): Sapling output description invalid"),
+                                      REJECT_INVALID, "bad-txns-sapling-output-description-invalid");
+            }
+        }
+
+        if (!librustzcash_sapling_final_check(
+            ctx,
+            tx.valueBalance,
+            &tx.bindingSig[0]
```
Comment (str4d) [inline src/main.cpp]: `tx.bindingSig.begin()`

Comment (str4d) [reply src/main.cpp]: Missing `sighashValue`.

**Comment by ebfull** [comment]
> Is it possible to write tests for this?

Right now there's no way for us to do this until we get some wallet PRs written.

**Comment by ebfull** [comment]
@zkbot try

**Comment by ebfull** [comment]
@zkbot try

**Comment by ebfull** [comment]
Assertions in Rust code will abort the process. Nothing in the Rust code should assert things that would otherwise be recoverable.

**Comment by bitcartel** [comment]
Is there a card up for "We must write tests in the PR that adds Sapling proving." ?

**Comment by Eirik0** [comment]
utACK

**Pull request merged**